rfd/0180-sso-as-mfa-method.md

```diff
@@ -0,0 +1,767 @@
+---
+authors: Brian Joerger (bjoerger@goteleport.com)
+state: draft
+---
+
+# RFD 180 - SSO MFA
+
+## Required Approvers
+
+* Engineering: @rosstimothy || @codingllama
+* Product: @xinding33 || @klizhentas
+
+## What
+
+Provide the ability to satisfy Teleport MFA checks via a 3rd party identity
+provider with SSO.
+
+Per-session MFA must be supported for all Teleport clients, while additional
+features will be included as a stretch goal, such as MFA for admin actions and
+moderated sessions. Ideally we will support SSO as a first class MFA method.
+
+## Why
+
+Delegating MFA checks to a registered IdP has several potential benefits:
+- Administrators gain the ability to configure and monitor all authentication
+directly through an IdP.
+- Teleport can integrate with custom MFA mechanisms and devices registered
+setup through an IdP.
```

Comment:
Thoughts from a product perspective: The same can be achieved by supporting external MFA provider by implementing MFA API provided by IdP such as [Okta verify factor API](https://developer.okta.com/docs/reference/api/authn/#verify-factor), [Duo auth API](https://duo.com/docs/authapi#/auth) etc.

Why support re-authentication over utilizing such API that lets us request just MFA? We will likely need to support external MFA provider some time in future.

Reply:
That's an interesting idea which I hadn't considered. However, integrating directly with IdP APIs could grow out of hand quickly. I'm not sure how many there are and which ones our customers currently use.

IIUC, each IdP should have settings to control how/which internal APIs are used during authentication - e.g. you can create an Okta SAML App that skips the username/password prompt for existing sessions and just prompts for MFA.

Creating a generalized implementation that supports OIDC and SAML instead of specific IdPs seems like a safer place to start to me. It's also the exact flow requested by a customer that uses an OIDC provider with a very customized MFA device which does not come into play directly. From what I understand authentication through this device is handled through through the IdP service app and would likely require several API calls to mimic the full flow. I would have to dive into the deeper details of their setup, which IMO shouldn't be necessary if we can just support the most generalizable use case and let customers design their own flows in their IdP.

Reply:
Okta does offer granular authentication policy and allows configuring number of auth factors and factor type, though I do not know how common is such feature in other IdPs. Either way, you are correct that solution with SSO will be the easiest and fastest to start with and enough for our use case.

To elaborate more on the implementation of the external MFA provider with - yeah, we would implement a base MFA API that passes MFA challenge and response back and forth between browser and configured MFA provider, among which one of such provider will be Teleport itself.